github/VisualStudio
# Warn when no origin remote

### What this PR does

If a repository doesn't have a remote named `origin`, the extension doesn't know which remote to use or even if it's dealing with a GitHub repository. This PR shows a call-out notification if a repository has remotes but none of them are named `origin`. If the used clicks on the notification title, the `GitHub` pane will open with an option to edit their remotes.

- Add `TippingService` for displaying call-out notifications on Visual Studio 2015, 2017 and 2019
- Display `NoRemoteOriginView` on `GitHub` pane when there are no remotes named `origin`
- Add `ShowRepositorySettingsRemotesAsync` to `ITeamExplorerServices`
- Prevent exception in `GetSessionInternal` when repository `CloneUrl` changes to null
- Add integration tests for `GitService` (that use a real repository)
- Add `HasRemotesButNoOrigin` property to `LocalRepositoryModel`
- Add `CantFindGitHubUrlForRepository` and `RepositoriesMustHaveRemoteOrigin` to `Resources`
   - These contain text for the notification and the pane view
- Add `TrackedRemoteName` to `BranchModel`
  - This property has tests but isn't currently used. I thought `NoRemoteOriginView` might suggest renaming the currently active remote (if any) to `origin`?

### Reviewers

- Is the wording of the call-out / pane view clear enough?

### What to expect

When a repository has at least one remote, but no remotes called "origin", the following call-out notification will appear.

![image](https://user-images.githubusercontent.com/11719160/48259758-e8bb5180-e410-11e8-8e22-92e22a093116.png)

Clicking the bonded title will open the `GitHub` pane, which will show the following message. Alternatively, clicking `Always ignore` will cause the call-out notification to _never_ show again.

![image](https://user-images.githubusercontent.com/11719160/49160363-8f16ba80-f31e-11e8-84c0-f0a8ca49d3d6.png)

Clicking the `Edit Remotes` button will open the `Repository Settings` page with the `Remotes` section visible. This lets the user rename one of their remotes to `origin`.

![image](https://user-images.githubusercontent.com/11719160/48488106-ec3b4800-e817-11e8-9323-8b00686c0577.png)

### To do

- [ ] Expand text on GitHub pane to guide user towards fix
- [ ] Add metrics

### How to test

1. Open solution from a repository on GitHub
2. From command line type
  `git remote rename origin porigin`
3. Expect call-out notification to appear
4. Click on title of call-out notification
5. Expect `GitHub` pane to appear with no "origin" message
6. Click on the `Edit Remotes` button
7. Expect `Repository Settings` page with the `Remotes` section visible
8. Edit the `porigin` rename and rename it to `origin`
9. Expect PR list to appear on `GitHub` pane

### Related

- #1727 Update the messaging in the GitHub panel when a solution is not open

Fixes #1729
